book/chapter-01-unix-sockets/01_unix_sockets_intro.md

```diff
@@ -0,0 +1,159 @@
+
+> So you use a stream socket when having information in order and intact is important. File transfer protocols are a good example here. You don't want to download some file with its contents randomly shuffled around and damaged!
+
+**Note well**: An important distinction to make with Unix datagram packets (compared to _network sockets_) is that the order of the packets are guaranteed, and the order will remain the same - but the analogy still holds!
+
+## An example
+We can create a quick example of how this works using python; although any language (including `ba/sh`) would be more than suffice here, although python has some advantages we can get into later (specifically - looking at contexts)!
+
+```python
+# server.py
+import socket
+
+with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
+    sock.bind("/tmp/docode.sock")
+    sock.listen()
+    print(f"Listening on {sock.getsockname()}")
+    conn, address = sock.accept()
+    with conn:
+        while True:
+            data = conn.recv(1024)
+            print(f"Got connection with message: {data}")
+            if not data:
+                break
+            conn.sendall(data)
+```
+
+If we run `python server.py` we can see the address we are listening on - but the program will appear to hang: it is awaiting a connection before proceeding!
+
+We can prove the kernel is listenting on this socket by creating another terminal instance, and running
+
+```sh
+# the -l flag filters out those to ensure only sockets that are listenting are returned
+# the -p flag shows the associated process
+netstat -lp 
+```
+
+This will return _every socket_ that is listening for connections - we can filter out all of them to just see our process with:
+
+```sh
+netstat -lp | grep docode
+```
+
+In this same terminal we can run `python client.py` - check the output of each terminal!
+
+```python
+# client.py
+import socket
+
+with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
+    sock.connect("/tmp/docode.sock")
+    sock.sendall(b"Hello, docode!")
+    data = sock.recv(1024)
+
+print(f"Received {data!r}")
+```
+
+Success! We should see that both the server and client process report the message!
+
+### How did that work?
+
+When calling 
+
+```python
+with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
+``` 
+
+we are using a special python keywork `with` - this is known as a [context manager](https://docs.python.org/3/library/stdtypes.html#typecontextmanager) in python, which allows us to run code before we enter the `with` block (defined by a `__enter__` method), and run code afterwards (defined by a `__exit__` block). This is usually done to make sure we clean up code or states that are easy to forget to write! Often this is used in contexts when interfacing with files or networking where we want to close ports or release resources back to the operating system.
+
```

Comment:
maybe we can code our own `__enter__` to demonstrate how this works in a short code example?

Reply:
Done ✅